pgwire: handle case where TypeHints != PreparedArguments

otan opened the pull request:
Release note (bug fix): Fix a panic where type hints mismatching
placeholder names cause a crash.

Resolves #59007.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59450)
<!-- Reviewable:end -->

rafiss commented on pkg/sql/conn_executor_prepare.go:
```diff
@@ -53,11 +53,20 @@ func (ex *connExecutor) execPrepare(
+				if i >= len(parseCmd.RawTypeHints) {
+					return retErr(pgerror.Newf(
+						pgcode.ProtocolViolation,
+						"expected type hint at position %d",
```
maybe we should try to match the error message that happens in the simple case:

https://github.com/cockroachdb/cockroach/blob/32bc171667b6e180fa09669802d45fc9edbe4a52/pkg/sql/conn_executor_prepare.go#L300

otan committed:
pgwire: handle case where TypeHints != PreparedArguments

Release note (bug fix): Fix a panic where type hints mismatching
placeholder names cause a crash.

otan commented:
bors r=rafiss
